### DISCUSSION TITLE
Writing QGIS pytests

### DISCUSSION THREAD

**Discussion by kaiguy23** [SMByC/AcATaMa#17] [Q&A]
Hi there!
I have recently been developing a plugin of my own for a very different use case, and I am just now getting around to writing test cases for it. Your plugin has been a great learning resource for me, and I have often returned to this repository for seeing how certain widgets are implemented and how to use the hundred of different qgis modules. I was wondering if you could answer a question for me regarding the pytests, which are run by the `run_tests.sh` file in `.docker/`.

I noticed that you are using the `qgis.testing.start_app()` to initially setup the application. My plugin interacts with several layers in a specific qgis project file and I was hoping to use this exact project file in the plugin pytests. For example, the plugin is responsible for changing certain sql filter expressions, adding new raster layers, getting/setting project variables, etc. While I read that it is possible to pass the `--project /path/to/my/project.qgs` as a `sys.argv`, the `QgsApplication` doesn't seem to be using the project.  I even created my own version of the `start_app()` function, and hardcoded the project as a `sys.argv`. Is there a way for the QgsApplication to actually use the project?

In addition, the `iface` object from `qgis.utils` (which I use for setting the active layer and window extent), doesn't seem to be available in the pytest mode. I saw the `pytest-qgis` creates a global `iface` variable, which replaces the actual `iface` with `None`. Is there a reason for this? Is is possible to actually use the original `iface` from `qgis.utils`?

I'd appreciate any advice you have to offer.
Thanks!

**Comment by XavierCLL** [comment]
Hi @kaiguy23 

Sure, but a while ago I wrote these tests I'll try to reply to you hahaha

Your first question, not sure if I understood well, but you want to load a project for the tests right?, you don't need to complicate changing that start_app() (and not recommended), you can load a project file inside a test function using directly the code for that `project.read('testdata/01_project.qgs')` check Qgis [documentation](https://docs.qgis.org/3.28/en/docs/pyqgis_developer_cookbook/loadproject.html), but honestly I have not tested it before but it should work.

Second, about the `iface`: remember that tests are running in a fake environment so they need a fake `iface` (and other fake things). For that, I use the plugin called [pytest-qgis](https://github.com/GispoCoding/pytest-qgis) (you need to setup it in your pytest conf file) the only that you need is to add `qgis_iface` in the argument of the test function and call `qgis_iface` as it like the real iface, for instance:

```python
def test_iface(qgis_iface):
    QgsProject.instance().addMapLayer(layer)
    qgis_iface.setActiveLayer(layer)
```

**Comment by kaiguy23** [comment]
Thanks for the fast reply! I really appreciate it. 

I was able to create a pytest fixture that loads the project each time using the `project.read` method. I am now having trouble loading the project, as the layers are from a database, and I am getting a layer unavailable error. I have added the server as an extra host in the docker compose file and have also mounted the qgis auth db file. However, I don’t think there is a way to directly use the auth db file in the `QgsProject`. I am currently setting the auth fb file in my modified version of the start_app() function, but the layers are still unavailable. 

As for the iface, my plugin often uses the `qgis.utils iface` directly, rather than inheriting it from the plugin itself. I could just try to pass the `iface` variable that is initially given to the plug-in to the other spots of the code, although that may be kind of sloppy. Is it possible to create a global iface variable that is replaced by the qgis_iface when the testing starts? Does that seem reasonable?
Thanks!

**Comment by XavierCLL** [comment]
No idea to load a layer from a database for the tests, if reading that layer from a database is part of your tests, ok, you could setup that database in the same docker container for that, otherwise, I recommended storage the layers as local files for tests. Another thing is to remember that you need to fix the paths of the files in the `QgsProject` file, based on the path of the files (data folder) in the docker container for your tests.

(iface): why do you want to do that? do you have any issue with the `qgis_iface` provided by the pytest-qgis plugin?

**Reply by kaiguy23** [reply]
So I am importing iface from qgis.utils in my plugin, and it is turned into a None for some reason (probably pytest_qgis). How can I replace the iface calls with the qgis_iface in the plugin source code?

**Reply by kaiguy23** [reply]
As for the project question, I now just have my plugin setup the required layers on startup.

**Reply by XavierCLL** [reply]
Weird, `qgis_iface` should return stubbed `QgsInterface`, something is missing in your setup, check the readme in the https://github.com/GispoCoding/pytest-qgis, I saw an issue related to the tests with this plugin for Qgis version < 3.18, try testing with versions higher than that. Another example (and more complex than mine tests) of pytests using this plugin for Qgis plugins is this one https://github.com/nlsfi/segment-reshape-qgis-plugin/tree/main/test

**Reply by kaiguy23** [reply]
I am using the docker qgis image final_3.28.4, so I think that should be sufficient. I'll checkout the example you sent, thanks.

**Reply by kaiguy23** [reply]
I just re-wrote my code to stop using the `qgis.utils iface` and just use the iface directly passed to the dockwidget everywhere instead. this seems to have solved the problem for now... I also left a comment on the pytest-qgis github to see if they can help.
Thanks again!
